### ISSUE TITLE
`undocumented-unsafe-blocks` is too strict

### ISSUE THREAD

**Issue by poliorcetics** [rust-lang/rust-clippy#9114]
### Summary

Since Rust 1.62.0 the `undocumented-unsafe-blocks` lint is too strict, failing to detect safety comments when they are just a line above the `unsafe` block

### Reproducer

I tried this code:

```rust
fn main() {
    let xs = vec![1, 2, 3, 4];
    let ys: Vec<_> = xs
        .into_iter()
        // Safety: this is always safe
        .map(|x| unsafe { unsafe_square(x) })
        .collect();
    dbg!(ys);
}

unsafe fn unsafe_square(a: u8) -> u16 {
    a as u16 * 2
}
```

Run

```
cargo clippy -- -W clippy::undocumented-unsafe-block
```

I expected to see this happen:

No warning

Instead, this happened:

```
warning: unsafe block missing a safety comment
 --> src/main.rs:6:18
  |
6 |         .map(|x| unsafe { unsafe_square(x) })
  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: requested on the command line with `-W clippy::undocumented-unsafe-blocks`
  = help: consider adding a safety comment on the preceding line
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#undocumented_unsafe_blocks
```

### Version

```text
rustc 1.62.0 (a8314ef7d 2022-06-27)
binary: rustc
commit-hash: a8314ef7d0ec7b75c336af2c9857bfaf43002bfc
commit-date: 2022-06-27
host: aarch64-apple-darwin
release: 1.62.0
LLVM version: 14.0.5
```
